File: src/main/java/dev/stashy/extrasounds/json/SoundEntrySerializer.java
package dev.stashy.extrasounds.json;

import com.google.gson.*;
import net.minecraft.client.sound.Sound;
import net.minecraft.client.sound.SoundEntry;

import java.lang.reflect.Type;
import java.util.Objects;

public class SoundEntrySerializer implements JsonSerializer<SoundEntry>
{
    @Override
    public JsonElement serialize(SoundEntry src, Type typeOfSrc, JsonSerializationContext context)
    {
        JsonObject o = new JsonObject();
        JsonArray sounds = new JsonArray();
        for (Sound snd : src.getSounds())
            sounds.add(context.serialize(snd));
        o.add("sounds", sounds);
        if (src.canReplace())
            o.addProperty("replace", src.canReplace());
        if (!Objects.equals(src.getSubtitle(), ""))
            o.addProperty("subtitle", src.getSubtitle());
        return o;
    }
}

File: src/main/kotlin/dev/stashy/extrasounds/json/SoundEntrySerializer.kt
package dev.stashy.extrasounds.json

import com.google.gson.*
import net.minecraft.client.sound.SoundEntry
import java.lang.reflect.Type

class SoundEntrySerializer : JsonSerializer<SoundEntry> {
    override fun serialize(src: SoundEntry, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        return JsonObject().also { obj ->
            obj.add("sounds", JsonArray().also { arr ->
                src.sounds.forEach { arr.add(context.serialize(it)) }
            })
            if (src.canReplace())
                obj.addProperty("replace", src.canReplace())
            if (src.subtitle != null)
                obj.addProperty("subtitle", src.subtitle)
        }
    }
}